### ISSUE TITLE
Open MPI v4.1.5 fails to establish OOB TCP communication when used with PMIx 4.2.3

### ISSUE THREAD

**Issue by dirteat** [open-mpi/ompi#11729]
Hi there,

version 4.1.5, on Mageia 9, from an rpm package that I do compile, linked with external openpmix (4.2.3). Tcp connect, with ssh working from and to all nodes.

This installation of openmpi works fine under slurm on a multiple nodes system. But a simple test using 
```shell
mpirun -np 2 -host "x,y" mytestprogram
```
fails. The logs is pretty uninformative. The best I can get by searching past issues is what follows
[ompi_info.log](https://github.com/open-mpi/ompi/files/11653992/ompi_info.log)
, still no obvious error located. Any advices to debug this welcome!

```shell
strace -f -e execve mpirun -np 2 --mca plm_base_verbose 10 -host "192.168.0.18,192.168.0.19" hostname

execve("/usr/bin/mpirun", ["mpirun", "-np", "2", "--mca", "plm_base_verbose", "10", "-host", "192.168.0.18,192.168.0.19", "hostname"], 0x7fff6567c7e8 /* 42 vars */) = 0
[galatea:12241] mca: base: components_register: registering framework plm components
[galatea:12241] mca: base: components_register: found loaded component isolated
[galatea:12241] mca: base: components_register: component isolated has no register or open function
[galatea:12241] mca: base: components_register: found loaded component rsh
[galatea:12241] mca: base: components_register: component rsh register function successful
[galatea:12241] mca: base: components_register: found loaded component slurm
[galatea:12241] mca: base: components_register: component slurm register function successful
[galatea:12241] mca: base: components_open: opening plm components
[galatea:12241] mca: base: components_open: found loaded component isolated
[galatea:12241] mca: base: components_open: component isolated open function successful
[galatea:12241] mca: base: components_open: found loaded component rsh
[galatea:12241] mca: base: components_open: component rsh open function successful
[galatea:12241] mca: base: components_open: found loaded component slurm
[galatea:12241] mca: base: components_open: component slurm open function successful
[galatea:12241] mca:base:select: Auto-selecting plm components
[galatea:12241] mca:base:select:(  plm) Querying component [isolated]
[galatea:12241] mca:base:select:(  plm) Query of component [isolated] set priority to 0
[galatea:12241] mca:base:select:(  plm) Querying component [rsh]
[galatea:12241] mca:base:select:(  plm) Query of component [rsh] set priority to 10
[galatea:12241] mca:base:select:(  plm) Querying component [slurm]
[galatea:12241] mca:base:select:(  plm) Selected component [rsh]
[galatea:12241] mca: base: close: component isolated closed
[galatea:12241] mca: base: close: unloading component isolated
[galatea:12241] mca: base: close: component slurm closed
[galatea:12241] mca: base: close: unloading component slurm
strace: Process 12242 attached
strace: Process 12243 attached
strace: Process 12244 attached
[galatea:12241] [[62120,0],0] plm:rsh: final template argv:
	/usr/bin/ssh <template>  orted -mca ess "env" -mca ess_base_jobid "4071096320" -mca ess_base_vpid "<template>" -mca ess_base_num_procs "2" -mca orte_node_regex "galatea,[3:192].168.0.19@0(2)" -mca orte_hnp_uri "4071096320.0;tcp://192.168.0.18:42961" --mca plm_base_verbose "10" -mca plm "rsh" --tree-spawn -mca routed "radix" -mca orte_parent_uri "4071096320.0;tcp://192.168.0.18:42961" -mca pmix "^s1,s2,cray,isolated"
strace: Process 12245 attached
[pid 12245] execve("/usr/bin/ssh", ["/usr/bin/ssh", "192.168.0.19", " orted", "-mca", "ess", "\"env\"", "-mca", "ess_base_jobid", "\"4071096320\"", "-mca", "ess_base_vpid", "1", "-mca", "ess_base_num_procs", "\"2\"", "-mca", "orte_node_regex", "\"galatea,[3:192].168.0.19@0(2)\"", "-mca", "orte_hnp_uri", "\"4071096320.0;tcp://192.168.0.18"..., "--mca", "plm_base_verbose", "\"10\"", "-mca", "plm", "\"rsh\"", "--tree-spawn", "-mca", "routed", "\"radix\"", "-mca", ...], 0x12f04f0 /* 46 vars */) = 0
strace: Process 12246 attached
[pid 12246] execve("/usr/bin/sh", ["sh", "-c", "/usr/bin/xauth  list unix:10.0 2"...], 0x21c18d0 /* 46 vars */) = 0
strace: Process 12247 attached
[pid 12247] execve("/usr/bin/xauth", ["/usr/bin/xauth", "list", "unix:10.0"], 0x11c49a0 /* 45 vars */) = 0
[pid 12247] +++ exited with 0 +++
[pid 12246] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=12247, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
[pid 12246] +++ exited with 0 +++
[pid 12245] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=12246, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
[lerna:11339] mca: base: components_register: registering framework plm components
[lerna:11339] mca: base: components_register: found loaded component rsh
[lerna:11339] mca: base: components_register: component rsh register function successful
[lerna:11339] mca: base: components_open: opening plm components
[lerna:11339] mca: base: components_open: found loaded component rsh
[lerna:11339] mca: base: components_open: component rsh open function successful
[lerna:11339] mca:base:select: Auto-selecting plm components
[lerna:11339] mca:base:select:(  plm) Querying component [rsh]
[lerna:11339] mca:base:select:(  plm) Query of component [rsh] set priority to 10
[lerna:11339] mca:base:select:(  plm) Selected component [rsh]
[lerna:11339] mca: base: close: component rsh closed
[lerna:11339] mca: base: close: unloading component rsh
[pid 12245] +++ exited with 1 +++
[pid 12241] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=12245, si_uid=1000, si_status=1, si_utime=7 /* 0.07 s */, si_stime=4 /* 0.04 s */} ---
--------------------------------------------------------------------------
ORTE was unable to reliably start one or more daemons.
This usually is caused by:

* not finding the required libraries and/or binaries on
  one or more nodes. Please check your PATH and LD_LIBRARY_PATH
  settings, or configure OMPI with --enable-orterun-prefix-by-default

* lack of authority to execute on one or more specified nodes.
  Please verify your allocation and authorities.

* the inability to write startup files into /tmp (--tmpdir/orte_tmpdir_base).
  Please check with your sys admin to determine the correct location to use.

*  compilation of the orted with dynamic libraries when static are required
  (e.g., on Cray). Please check your configure cmd line and consider using
  one of the contrib/platform definitions for your system type.

* an inability to create a connection back to mpirun due to a
  lack of common network interfaces and/or no route found between
  them. Please check network connectivity (including firewalls
  and network routing requirements).
--------------------------------------------------------------------------
[pid 12242] +++ exited with 0 +++
[pid 12243] +++ exited with 0 +++
[galatea:12241] mca: base: close: component rsh closed
[galatea:12241] mca: base: close: unloading component rsh
[pid 12244] +++ exited with 0 +++
+++ exited with 1 +++
```

**Comment by ggouaillardet** [comment]
What if you use the absolute path to `mpirun`?

**Comment by dirteat** [comment]
Woo, that is fast answer!

I've just tried:

```shell
strace -f -e execve /usr/bin/mpirun -np 2 --mca plm_base_verbose 10 -host 192.168.0.18,192.168.0.1 hostname
```
and it produces the same output.

Just a comment, I assume that in the output of strace, the "<template>" would be actually replaced by some real "template"?


NB: To be sure I am not screwing it up:
```shell
which mpirun
/usr/bin/mpirun

mpirun --version
mpirun (Open MPI) 4.1.5

Report bugs to http://www.open-mpi.org/community/help/
```

**Comment by ggouaillardet** [comment]
>
> strace is unlikely to help you and could break things if setuid bit
programs are involved.

What if you simply?
ssh other_node_ip orted

**Comment by dirteat** [comment]
```shell
ssh 192.168.0.19 orted
[lerna:35736] [[INVALID],INVALID] ORTE_ERROR_LOG: Not found in file ess_env_module.c at line 135
[lerna:35736] [[INVALID],INVALID] ORTE_ERROR_LOG: Bad parameter in file util/session_dir.c at line 107
[lerna:35736] [[INVALID],INVALID] ORTE_ERROR_LOG: Bad parameter in file util/session_dir.c at line 346
[lerna:35736] [[INVALID],INVALID] ORTE_ERROR_LOG: Bad parameter in file base/ess_base_std_orted.c at line 264
--------------------------------------------------------------------------
It looks like orte_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during orte_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  orte_session_dir failed
  --> Returned value Bad parameter (-5) instead of ORTE_SUCCESS
--------------------------------------------------------------------------
```

And if I do the same on the current node I am getting the same result:

```shell
orted

[galatea:37161] [[INVALID],INVALID] ORTE_ERROR_LOG: Not found in file ess_env_module.c at line 135
[galatea:37161] [[INVALID],INVALID] ORTE_ERROR_LOG: Bad parameter in file util/session_dir.c at line 107
[galatea:37161] [[INVALID],INVALID] ORTE_ERROR_LOG: Bad parameter in file util/session_dir.c at line 346
[galatea:37161] [[INVALID],INVALID] ORTE_ERROR_LOG: Bad parameter in file base/ess_base_std_orted.c at line 264
--------------------------------------------------------------------------
It looks like orte_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during orte_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  orte_session_dir failed
  --> Returned value Bad parameter (-5) instead of ORTE_SUCCESS
--------------------------------------------------------------------------
```

But, doing, on the current node (192.168.0.18, galatea)
```shell
mpirun -np 1 -host 192.168.0.18 hostname
galatea
```

works.

**Comment by ggouaillardet** [comment]
Ok, so this is not a simple "library not found" library...

What if you

`/usr/bin/mpirun --mca oob_base_verbose 100 --host ip_of_the_other_node -np
1 hostname`

FWIW, Open MPI does not spawn an `orted` daemon when a single node is
involved.

BTW, make sure all orted daemons are identical on all your nodes.

**Comment by dirteat** [comment]
Here we go:

```shell

[chris@galatea ~]$ /usr/bin/mpirun --mca oob_base_verbose 100 --host 192.168.0.19 -np 1 hostname
[galatea:05518] mca: base: components_register: registering framework oob components
[galatea:05518] mca: base: components_register: found loaded component tcp
[galatea:05518] mca: base: components_register: component tcp register function successful
[galatea:05518] mca: base: components_open: opening oob components
[galatea:05518] mca: base: components_open: found loaded component tcp
[galatea:05518] mca: base: components_open: component tcp open function successful
[galatea:05518] mca:oob:select: checking available component tcp
[galatea:05518] mca:oob:select: Querying component [tcp]
[galatea:05518] oob:tcp: component_available called
[galatea:05518] WORKING INTERFACE 1 KERNEL INDEX 1 FAMILY: V4
[galatea:05518] [[51447,0],0] oob:tcp:init rejecting loopback interface lo
[galatea:05518] WORKING INTERFACE 2 KERNEL INDEX 6 FAMILY: V4
[galatea:05518] [[51447,0],0] oob:tcp:init adding 192.168.0.18 to our list of V4 connections
[galatea:05518] [[51447,0],0] TCP STARTUP
[galatea:05518] [[51447,0],0] attempting to bind to IPv4 port 0
[galatea:05518] [[51447,0],0] assigned IPv4 port 33931
[galatea:05518] mca:oob:select: Adding component to end
[galatea:05518] mca:oob:select: Found 1 active transports
[galatea:05518] [[51447,0],0]: get transports
[galatea:05518] [[51447,0],0]:get transports for component tcp
[lerna:05505] mca: base: components_register: registering framework oob components
[lerna:05505] mca: base: components_register: found loaded component tcp
[lerna:05505] mca: base: components_register: component tcp register function successful
[lerna:05505] mca: base: components_open: opening oob components
[lerna:05505] mca: base: components_open: found loaded component tcp
[lerna:05505] mca: base: components_open: component tcp open function successful
[lerna:05505] mca:oob:select: checking available component tcp
[lerna:05505] mca:oob:select: Querying component [tcp]
[lerna:05505] oob:tcp: component_available called
[lerna:05505] WORKING INTERFACE 1 KERNEL INDEX 1 FAMILY: V4
[lerna:05505] [[51447,0],1] oob:tcp:init rejecting loopback interface lo
[lerna:05505] WORKING INTERFACE 2 KERNEL INDEX 6 FAMILY: V4
[lerna:05505] [[51447,0],1] oob:tcp:init adding 192.168.0.19 to our list of V4 connections
[lerna:05505] [[51447,0],1] TCP STARTUP
[lerna:05505] [[51447,0],1] attempting to bind to IPv4 port 0
[lerna:05505] [[51447,0],1] assigned IPv4 port 40113
[lerna:05505] mca:oob:select: Adding component to end
[lerna:05505] mca:oob:select: Found 1 active transports
[lerna:05505] [[51447,0],1]: get transports
[lerna:05505] [[51447,0],1]:get transports for component tcp
[lerna:05505] [[51447,0],1] OOB_SEND: rml_oob_send.c:265
[lerna:05505] [[51447,0],1] oob:base:send to target [[51447,0],0] - attempt 0
[lerna:05505] [[51447,0],1] oob:base:send unknown peer [[51447,0],0]
[lerna:05505] [[51447,0],1] oob:tcp:send_nb to peer [[51447,0],0]:63 seq = -1
[lerna:05505] [[51447,0],1]:[oob_tcp.c:188] processing send to peer [[51447,0],0]:63 seq_num = -1 hop [[51447,0],0] unknown
[lerna:05505] [[51447,0],1]:[oob_tcp.c:191] post no route to [[51447,0],0]
[lerna:05505] [[51447,0],1] OOB_SEND: rml_oob_send.c:265
[lerna:05505] [[51447,0],1] tcp:no route called for peer [[51447,0],0]
[lerna:05505] [[51447,0],1] OOB_SEND: oob_tcp_component.c:1123
[lerna:05505] [[51447,0],1] oob:base:send to target [[51447,0],0] - attempt 0
[lerna:05505] [[51447,0],1] oob:base:send unknown peer [[51447,0],0]
[lerna:05505] [[51447,0],1] oob:tcp:send_nb to peer [[51447,0],0]:10 seq = -1
[lerna:05505] [[51447,0],1]:[oob_tcp.c:188] processing send to peer [[51447,0],0]:10 seq_num = -1 hop [[51447,0],0] unknown
[lerna:05505] [[51447,0],1]:[oob_tcp.c:191] post no route to [[51447,0],0]
[lerna:05505] [[51447,0],1] oob:base:send to target [[51447,0],0] - attempt 1
[lerna:05505] [[51447,0],1] oob:base:send known transport for peer [[51447,0],0]
[lerna:05505] [[51447,0],1] oob:tcp:send_nb to peer [[51447,0],0]:63 seq = -1
[lerna:05505] [[51447,0],1]:[oob_tcp.c:188] processing send to peer [[51447,0],0]:63 seq_num = -1 hop [[51447,0],0] unknown
[lerna:05505] [[51447,0],1]:[oob_tcp.c:191] post no route to [[51447,0],0]
[lerna:05505] [[51447,0],1] tcp:no route called for peer [[51447,0],0]
[lerna:05505] [[51447,0],1] OOB_SEND: oob_tcp_component.c:1123
[lerna:05505] [[51447,0],1] tcp:no route called for peer [[51447,0],0]
[lerna:05505] [[51447,0],1] OOB_SEND: oob_tcp_component.c:1123
[lerna:05505] [[51447,0],1] oob:base:send to target [[51447,0],0] - attempt 1
[lerna:05505] [[51447,0],1] oob:base:send unknown peer [[51447,0],0]
[lerna:05505] [[51447,0],1] oob:tcp:send_nb to peer [[51447,0],0]:10 seq = -1
[lerna:05505] [[51447,0],1]:[oob_tcp.c:188] processing send to peer [[51447,0],0]:10 seq_num = -1 hop [[51447,0],0] unknown
[lerna:05505] [[51447,0],1]:[oob_tcp.c:191] post no route to [[51447,0],0]
[lerna:05505] [[51447,0],1] oob:base:send to target [[51447,0],0] - attempt 2
[lerna:05505] [[51447,0],1] oob:base:send known transport for peer [[51447,0],0]
[lerna:05505] [[51447,0],1] oob:tcp:send_nb to peer [[51447,0],0]:63 seq = -1
[lerna:05505] [[51447,0],1]:[oob_tcp.c:188] processing send to peer [[51447,0],0]:63 seq_num = -1 hop [[51447,0],0] unknown
[lerna:05505] [[51447,0],1]:[oob_tcp.c:191] post no route to [[51447,0],0]
[lerna:05505] [[51447,0],1] tcp:no route called for peer [[51447,0],0]
[lerna:05505] [[51447,0],1] OOB_SEND: oob_tcp_component.c:1123
[lerna:05505] [[51447,0],1] tcp:no route called for peer [[51447,0],0]
[lerna:05505] [[51447,0],1] OOB_SEND: oob_tcp_component.c:1123
[lerna:05505] [[51447,0],1] oob:base:send to target [[51447,0],0] - attempt 2
[lerna:05505] [[51447,0],1] oob:base:send unknown peer [[51447,0],0]
[lerna:05505] [[51447,0],1] oob:tcp:send_nb to peer [[51447,0],0]:10 seq = -1
[lerna:05505] [[51447,0],1]:[oob_tcp.c:188] processing send to peer [[51447,0],0]:10 seq_num = -1 hop [[51447,0],0] unknown
[lerna:05505] [[51447,0],1]:[oob_tcp.c:191] post no route to [[51447,0],0]
[lerna:05505] [[51447,0],1] oob:base:send to target [[51447,0],0] - attempt 3
[lerna:05505] [[51447,0],1] TCP SHUTDOWN
[lerna:05505] no hnp or not active
[lerna:05505] [[51447,0],1] TCP SHUTDOWN done
[lerna:05505] mca: base: close: component tcp closed
[lerna:05505] mca: base: close: unloading component tcp
--------------------------------------------------------------------------
ORTE was unable to reliably start one or more daemons.
This usually is caused by:

* not finding the required libraries and/or binaries on
  one or more nodes. Please check your PATH and LD_LIBRARY_PATH
  settings, or configure OMPI with --enable-orterun-prefix-by-default

* lack of authority to execute on one or more specified nodes.
  Please verify your allocation and authorities.

* the inability to write startup files into /tmp (--tmpdir/orte_tmpdir_base).
  Please check with your sys admin to determine the correct location to use.

*  compilation of the orted with dynamic libraries when static are required
  (e.g., on Cray). Please check your configure cmd line and consider using
  one of the contrib/platform definitions for your system type.

* an inability to create a connection back to mpirun due to a
  lack of common network interfaces and/or no route found between
  them. Please check network connectivity (including firewalls
  and network routing requirements).
--------------------------------------------------------------------------
[galatea:05518] [[51447,0],0] TCP SHUTDOWN
[galatea:05518] [[51447,0],0] TCP SHUTDOWN done
[galatea:05518] mca: base: close: component tcp closed
[galatea:05518] mca: base: close: unloading component tcp
```

PS: I've checked the orted daemons are the same (and diff of the binaries output nothing).

**Comment by jsquyres** [comment]
It looks like the launched orted is unable to call back to the mpirun -- it somehow concludes that there is no route.

Can you send information about the network you are using, and the output from `ip addr` and `ip route` on both the machine on which you invoked `mpirun` and 192.168.0.19?

Do you have any firewalls between the two?  Is host firewalling enabled on either machine?

**Comment by jsquyres** [comment]
@dirteat I am still investigating this; I don't see an obvious cause for the issue.  Could you do me a favor and try this out using a non-bonded interface?  It *shouldn't* make a difference, but I'd like to eliminate that as a possibility -- especially since Open MPI v4.1.x. is pretty mature at this point, and we haven't seen a failure like this before.

**Comment by dirteat** [comment]
No pb, thank you for all your feedback.

I have already tested that under non-bonded interfaces (between two desktops actually), same errors, failure in establishing the tcp communication. I should may be mention that all these machines run very recent kernels (version 6.3.6).

In the meanwhile I am testing a build without external pmix nor hwloc to see how it goes.

**Comment by dirteat** [comment]
Damned... That is working. My pb seems to be building against external hwloc/pmix, which should not be a problem I suspect?

I should check who is the guilty guy (pmix or hwloc). I see during the build that openmpi bundles pmixv3, while I was linking to pmixv4.

**Comment by jsquyres** [comment]
Ah!  That's the key info: external, new PMIx.

When I use an external PMIx (v4.2.3), I am able to replicate your behavior.  Let me re-title this issue and continue to investigate.

**Comment by rhc54** [comment]
See the patch from Gilles - it fixes the issues.

**Comment by jsquyres** [comment]
I did some more testing and talked with @rhc54 in Slack about this.

* I confirm that PMIx 4.2.3 does not work with Open MPI 4.1.x.
* However, PMIx <= v4.2.2 *does* work.
* Apparently OMPI v4.1.x is using some internal PMIx functions/macros/whatever, and those changed in PMIx v4.2.3 (and likely won't be changing back).  OMPI PR #11472 should fix this issue by updating OMPI v4.1.x to use the proper functions/macros.

**Comment by dirteat** [comment]
Thank you, you rock! Patch applied and tested, works again with external pmix 4.2.3!

Cheers,
Chris.

**Comment by jsquyres** [comment]
Fix (#11472) has been applied to the v4.1.x branch; closing this issue.
